### DISCUSSION TITLE
Why is my code returning an empty Rule?

### DISCUSSION THREAD

**Discussion by luisvgs** [pest-parser/pest#624] [General]
Hi, I'm currently playing around with the library, and for testing purposes my grammar only consists of a String. 

```
program = _{ SOI ~ "\n"* ~ (Str ~ "\n"+) * ~ Str ? ~ EOI }
Str = @{ "'" ~ ( "''" | (!"'" ~ ANY) )* ~ "'" }
```

With the following code:
```
extern crate pest;
#[macro_use]
extern crate pest_derive;

use pest::error::Error;
use pest::iterators::Pair;
use pest::Parser;

#[derive(Parser)]
#[grammar = "conway.pest"]
struct ConwayParser;

#[derive(PartialEq, Debug, Clone)]
pub enum AstNode {
    Str(String),
}

pub fn string_parser(string_node: Pair<Rule>) -> AstNode {
    let mut string = String::new();

    for node in string_node.clone().into_inner() {
        match node.as_rule() {
            Rule::Str => string = String::from(node.to_string()),
            _ => unreachable!()
        }
    }
    AstNode::Str(string)
}
pub fn build_ast_from_expr(pair: pest::iterators::Pair<Rule>) -> AstNode {
    match pair.as_rule() {
        Rule::Str => string_parser(pair),
        unknown => panic!("Unknown expr: {:?}", unknown),
    }
}

pub fn parse(source: &str) -> Result<Vec<AstNode>, Error<Rule>> {
    let mut ast = vec![];

    let pairs = ConwayParser::parse(Rule::program, source)?;
    for pair in pairs {
        match pair.as_rule() {
            Rule::Str => {
                ast.push(build_ast_from_expr(pair));
            }
            _ => {}
        }
    }

    Ok(ast)
}

fn main() {
    let unparsed_file = std::fs::read_to_string("conway.cy")
        .expect("cannot read conway file");
    let astnode = parse(&unparsed_file).expect("unsuccessful parse");
    println!("{:?}", &astnode);
}
```

However, instead of returning `[Str("Hello world")]` when I run my code it returns just an empty Rule:

```
[Str("")]
```

Also, although the code runs, rust-analyzer complains that the proc-macro server crashed, not allowing the Parser macro to expand. Whether or not I enable such feature, the warning as well as the code output is the same. I suspect this might be the root of the problem? Any advice would be deeply appreciated.

**Comment by rob-mur** [comment]
I've created a repo which replicates your issue and then gives a hint of where to go. [https://github.com/rob-mur/pest-521](url)

If you run this code you'll see that it prints out the string node you are trying to parse, which you will see has no inners. This means that string_node.clone().into_inner() is empty and so line 21-26 are essentially not called, leading to the empty string being returned.

After you work it out, I recommend using [https://crates.io/crates/pest_consume](url) to avoid this kinds of errors in future. Also, reading [https://pest.rs/book/grammars/syntax.html](url) to try and simplify your grammar as there is built in functionality to deal with whitespace.

Shout if you are still stuck.
